api/api_test.go

```diff
@@ -72,6 +72,135 @@ func TestAddFilterFailure(t *testing.T) {
+	})
+
+	Convey("Successfully create a dimension with a request body but no options", t, func() {
+		reader := strings.NewReader("{}")
+		r, err := http.NewRequest("POST", "http://localhost:22100/filters/12345678/dimensions/age", reader)
+		So(err, ShouldBeNil)
+		w := httptest.NewRecorder()
+		api := CreateFilterAPI(host, mux.NewRouter(), &mocks.DataStore{}, &mocks.FilterJob{})
+		api.router.ServeHTTP(w, r)
+		So(w.Code, ShouldEqual, http.StatusCreated)
+	})
+
+	Convey("Successfully create a dimension with options", t, func() {
+		reader := strings.NewReader("{\"values\":[\"22\",\"17\"]}")
+		r, err := http.NewRequest("POST", "http://localhost:22100/filters/12345678/dimensions/age", reader)
+		So(err, ShouldBeNil)
+		w := httptest.NewRecorder()
+		api := CreateFilterAPI(host, mux.NewRouter(), &mocks.DataStore{}, &mocks.FilterJob{})
+		api.router.ServeHTTP(w, r)
+		So(w.Code, ShouldEqual, http.StatusCreated)
+	})
+}
+
+func TestAddFilterJobDimensionFailure(t *testing.T) {
+	t.Parallel()
+	Convey("When no data store is available, an internal error is returned", t, func() {
+		reader := strings.NewReader("{\"values\":[\"22\",\"17\"]}")
+		r, err := http.NewRequest("POST", "http://localhost:22100/filters/12345678/dimensions/age", reader)
+		So(err, ShouldBeNil)
+
+		w := httptest.NewRecorder()
+
+		api := CreateFilterAPI(host, mux.NewRouter(), &mocks.DataStore{InternalError: true}, &mocks.FilterJob{})
+		api.router.ServeHTTP(w, r)
+		So(w.Code, ShouldEqual, http.StatusInternalServerError)
+	})
+
+	Convey("When an invalid json message is sent, a bad request is returned", t, func() {
+		reader := strings.NewReader("{")
+		r, err := http.NewRequest("POST", "http://localhost:22100/filters/12345678/dimensions/age", reader)
+		So(err, ShouldBeNil)
+		w := httptest.NewRecorder()
+		api := CreateFilterAPI(host, mux.NewRouter(), &mocks.DataStore{}, &mocks.FilterJob{})
+		api.router.ServeHTTP(w, r)
+		So(w.Code, ShouldEqual, http.StatusBadRequest)
+	})
+
+	Convey("When a filter job does not exist, a not found is returned", t, func() {
+		reader := strings.NewReader("{\"values\":[\"22\",\"17\"]}")
+		r, err := http.NewRequest("POST", "http://localhost:22100/filters/12345678/dimensions/age", reader)
+		So(err, ShouldBeNil)
+		w := httptest.NewRecorder()
+		api := CreateFilterAPI(host, mux.NewRouter(), &mocks.DataStore{NotFound: true}, &mocks.FilterJob{})
+		api.router.ServeHTTP(w, r)
+		So(w.Code, ShouldEqual, http.StatusNotFound)
+	})
+
+	Convey("When a filter job has been submitted, a forbidden status is returned", t, func() {
+		reader := strings.NewReader("{\"values\":[\"22\",\"17\"]}")
+		r, err := http.NewRequest("POST", "http://localhost:22100/filters/12345678/dimensions/age", reader)
+		So(err, ShouldBeNil)
+		w := httptest.NewRecorder()
+		api := CreateFilterAPI(host, mux.NewRouter(), &mocks.DataStore{Forbidden: true}, &mocks.FilterJob{})
+		api.router.ServeHTTP(w, r)
+		So(w.Code, ShouldEqual, http.StatusForbidden)
```

Comment:
Instead of return a forbidden status, could we return lock status (423)? This comes from an extension (RFC 4918) on top of the HTTP protocol.